### PR TITLE
Bugfix verifier after gm update 3.3.0

### PR DESCRIPTION
**Pull request by kbeker**
No issue for that. Still missing: https://github.com/golemfactory/concent/issues/1127 and must be added before merging this PR.

### CODE REVIEW

File: concent_api/api-e2e-additional-verification-test.py [hunk 1]
```diff
@@ -144,8 +147,14 @@ def test_case_1_test_for_positive_case(cluster_consts: ProtocolConstants, cluste
...
         price=10000,
+        render_parameters=dict(
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/api-e2e-additional-verification-test.py]: Good point, done :)

File: concent_api/api-e2e-additional-verification-test.py [hunk 2]
```diff
@@ -417,84 +452,21 @@ def test_case_6_test_without_script_src_in(cluster_consts: ProtocolConstants, cl
...
-        subtask_results_verify.subtask_results_rejected.timestamp,
-        cluster_consts.additional_verification_call_time,
-        subtask_results_verify.subtask_results_rejected.report_computed_task.size,
-        cluster_consts.minimum_upload_rate,
-    ) - current_time
-    print(f"Going to sleep for {sleep_time} secs...")
-    time.sleep(
-        sleep_time
-    )
+    # Deleting parameters which are necessary for proper blender work
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/api-e2e-additional-verification-test.py]: Done

File: concent_api/api-e2e-additional-verification-test.py [hunk 3]
```diff
@@ -386,15 +423,13 @@ def test_case_5_test_requestor_status_account_negative(cluster_consts: ProtocolC
 
 
 @count_fails
-def test_case_6_test_without_script_src_in(cluster_consts: ProtocolConstants, cluster_url: str) -> None:  # pylint: disable=unused-argument
+def test_case_6_test_without_script_parameters(cluster_consts: ProtocolConstants, cluster_url: str) -> None:  # pylint: disable=unused-argument
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/api-e2e-additional-verification-test.py]: Done

File: concent_api/api_testing_common.py [hunk 4]
```diff
@@ -383,7 +383,7 @@ def create_signed_task_to_compute(
     price: int=1,
     size: int=1,
     package_hash: str='sha1:57786d92d1a6f7eaaba1c984db5e108c68b03f0d',
-    script_src: Optional[str]=None,
+    render_parameters: Optional[Dict[str, Union[bool, int, List[float]]]]=None,
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/api_testing_common.py]: Yea, you're right, it is too complicated and when I see at this next day I see that this is over-engineering adding this typing and I left now `Optional[Dict[str, Any]]` and I will change it everywhere where I see that it is too complicated.

File: concent_api/api_testing_common.py [hunk 5]
```diff
@@ -401,11 +401,17 @@ def create_signed_task_to_compute(
                 'output_format': 'png',
                 'scene_file': '/golem/resources/golem-header-light.blend',
                 'frames': [1],
-                'script_src': script_src,
+                'resolution': render_parameters.get('resolution') if render_parameters is not None else [400, 400],
+                'use_compositing': render_parameters.get('use_compositing') if render_parameters is not None else False,
+                'samples': render_parameters.get('samples') if render_parameters is not None else 0,
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/api_testing_common.py]: Yes, when we set `sample=0` it means that Blenders gets it's parameters directly from scene

File: concent_api/conductor/migrations/0013_auto_20190325_1258.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('samples', models.IntegerField()),
+                ('use_compositing', models.BooleanField()),
+            ],
+        ),
+        migrations.CreateModel(
+            name='ImageBorders',
+            fields=[
+                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
+                ('minimum_border_edge', models.DecimalField(decimal_places=4, max_digits=5)),
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/conductor/migrations/0013_auto_20190325_1258.py]: After talk with @Jakub89 he enlightened me that it could be more than 4 decimal_places and I think that I will add more then 4.

File: concent_api/conductor/service.py [hunk 7]
```diff
@@ -75,7 +80,7 @@ def store_verification_request_and_blender_subtask_definition(
     output_format: str,
     scene_file: str,
     verification_deadline: int,
-    blender_crop_script: Optional[str],
+    blender_parameters: Dict[str, Union[int, List[float], bool]],
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/conductor/service.py]: Done:)

File: concent_api/conductor/tasks.py [hunk 8]
```diff
@@ -39,7 +41,7 @@ def blender_verification_request(
     scene_file: str,
     verification_deadline: int,
     frames: List[int],
-    blender_crop_script: Optional[str],
+    blender_crop_script_parameters: Dict[str, Union[int, List[float], bool]],
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/conductor/tasks.py]: Doneee

File: concent_api/core/tests/test_unit_validation.py [hunk 9]
```diff
@@ -441,3 +447,211 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+    )  # pylint: disable=no-self-use
+    def test_that_invalid_resolution_value_raise_exception(self, resolution):
+        with pytest.raises(ConcentValidationError) as exception_wrapper:
+            validate_resolution(resolution)
+        assert_that(exception_wrapper.value.error_code).is_equal_to(ErrorCode.MESSAGE_INVALID)
+
+    @pytest.mark.parametrize(
+        'resolution', [
+            [1, 1],
+            [100000000, 100000000],
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: I couldn't find their maximal supported resolution. I just add it as check that with big resolutions this validator won't crash :)

File: concent_api/core/tests/test_unit_validation.py [hunk 10]
```diff
@@ -441,3 +447,211 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+        ]
+    )
+    def test_that_valid_crops_value_doesnt_raise_exception(self, borders_x, borders_y):
+        try:
+            validate_crops([self.create_crop_dict(borders_x, borders_y)])
+        except Exception as exception:  # pylint: disable=broad-except
+            assert False, f"Unexpected exception has been raised: {str(exception)}"
+
+    @pytest.mark.parametrize(
+        ['borders_x', 'borders_y'], [
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: There was only test on `beginning = end` but now I added also when `beginning > end`

File: concent_api/core/tests/utils.py [hunk 11]
```diff
@@ -248,6 +248,7 @@ def _get_deserialized_task_to_compute(
         signer_private_key: Optional[bytes] = None,
         size: int = 1,
         frames: Optional[List[int]] = None,
+        render_parameters: Optional[Dict[str, Union[int, List[float], bool]]] = None
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/utils.py]: Done:)

File: concent_api/core/utils.py [hunk 12]
```diff
@@ -200,3 +204,13 @@ def get_latest_existing_block_at(self, timestamp: int) -> Block:
...
         block = self._sci.get_block_by_number(highest)
         return block
+
+
+def extract_blender_parameters_from_compute_task_def(extra_data: Any) -> Dict[str, Union[List[float], bool, int]]:
+    return dict(
+        resolution=extra_data['resolution'],
+        samples=extra_data['samples'],
+        use_compositing=extra_data['use_compositing'],
+        borders_x=extra_data['crops'][0]['borders_x'],
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/utils.py]: It is validate in our validation and we only agree when `crops` has only one dict inside

File: concent_api/verifier/tests/test_utils.py [hunk 13]
```diff
@@ -16,7 +16,7 @@
 from core.tests.utils import generate_uuid_for_tests
 from verifier.exceptions import VerificationError
 from verifier.exceptions import VerificationMismatch
-from verifier.utils import get_files_list_from_archive
+from verifier.utils import get_files_list_from_archive, generate_blender_script
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/verifier/tests/test_utils.py]: Crap, I thought that I checked all imports... Done

File: concent_api/verifier/utils.py [hunk 14]
```diff
@@ -439,14 +433,29 @@ def render_images_by_frames(
...
-        render_image(frame_number, output_format, scene_file, subtask_id, verification_deadline, blender_crop_script)
+        render_image(frame_number, output_format, scene_file, subtask_id, verification_deadline, blender_crop_script_parameters)
         blender_out_file_name = generate_full_blender_output_file_name(scene_file, frame_number, output_format)
+        log(
+            logger,
+            f'blender_out_file_name: {blender_out_file_name}',
+            logging_level=LoggingLevel.WARNING,
+        )
         blender_output_file_name_list.append(blender_out_file_name)
+        log(
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/verifier/utils.py]: This was actually my logs for debug, but I add new logs to log more things in verifier

File: concent_api/verifier/utils.py [hunk 15]
```diff
@@ -439,14 +433,29 @@ def render_images_by_frames(
...
+            logger,
+            f'blender_output_file_name_list: {blender_output_file_name_list}',
+            logging_level=LoggingLevel.WARNING,
+        )
         parsed_files_to_compare[frame_number].append(blender_out_file_name)
+        log(
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/verifier/utils.py]: Same as https://github.com/golemfactory/concent/pull/1128#discussion_r269054721

File: concent_api/api-e2e-additional-verification-test.py [hunk 16]
```diff
@@ -417,84 +431,21 @@ def test_case_6_test_without_script_src_in(cluster_consts: ProtocolConstants, cl
...
-        cluster_url,
-        'receive',
-        sci_base.requestor_private_key,
-        CONCENT_PUBLIC_KEY,
-        create_client_auth_message(sci_base.requestor_private_key, sci_base.requestor_public_key, CONCENT_PUBLIC_KEY),
-        expected_status=200,
-        expected_message_type=message.concents.SubtaskResultsSettled,
-        expected_content_type='application/octet-stream',
-    )
+    # Setting invalid parameters which are necessary for proper blender work
```
Comment (dybi) [inline concent_api/api-e2e-additional-verification-test.py]: Hm... I would rephrase this sentence:
```suggestion
    # Setting parameters which are necessary for proper blender work to invalid values
```

Comment (kbeker) [reply concent_api/api-e2e-additional-verification-test.py]: Done

**Comment by kbeker** [comment]
I have squashed all fixups and waiting for #1129

File: concent_api/conductor/models.py [hunk 17]
```diff
@@ -70,6 +72,27 @@ class ResultTransferRequest(Model):
...
+    y_dimension = IntegerField()
+
+
+class BlenderCropScriptParameters(Model):
+
+    resolution = OneToOneField(ImageResolution, related_name='blender_crop_script_parameters')
+    samples = IntegerField()
+    use_compositing = BooleanField()
+    borders_y = OneToOneField(ImageBorders, related_name='blender_crop_script_parameters_as_borders_y')
+    borders_x = OneToOneField(ImageBorders, related_name='blender_crop_script_parameters_as_borders_x')
```
Comment (cameel) [inline concent_api/conductor/models.py]: This is not wrong but I'd personally just put the integers here instead of adding a relation:

 I.e.
``` python
resolution_x = IntegerField()
resolution_y = IntegerField()

borders_x_min = DecimalField(max_digits=10, decimal_places=9)
borders_x_max = DecimalField(max_digits=10, decimal_places=9)
borders_y_min = DecimalField(max_digits=10, decimal_places=9)
borders_y_max = DecimalField(max_digits=10, decimal_places=9)
```

File: concent_api/conductor/models.py [hunk 18]
```diff
@@ -94,6 +117,9 @@ class OutputFormat(ChoiceEnum):
     # Source code of the Python script to be executed by Blender.
     blender_crop_script = TextField(blank=True, null=True)
 
+    # Parameters needed to generate Python script to be executed by Blender.
+    blender_crop_script_parameters = OneToOneField(BlenderCropScriptParameters, related_name='blender_subtask_definition', blank=True, null=True)
```
Comment (cameel) [inline concent_api/conductor/models.py]: Same here. I'd put fields from `BlenderCropScriptParameters` directly here.

File: concent_api/conductor/service.py [hunk 19]
```diff
@@ -113,3 +119,38 @@ def store_frames(
...
+    )
+    borders_x.full_clean()
+    borders_x.save()
+
+    borders_y = ImageBorders(
+        minimum_border_edge=blender_parameters['borders_y'][0],
+        maximum_border_edge=blender_parameters['borders_y'][1],
+    )
+    borders_y.full_clean()
+    borders_y.save()
```
Comment (cameel) [inline concent_api/conductor/service.py]: You don't need to save these models separately. Django will do it automatically when you save `blender_crop_script_parameters`.

`blender_crop_script_parameters.full_clean()` will probably also run `full_clean()` on the related objects too, though I'm not 100% sure about that.

File: concent_api/conductor/utils.py [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+        blender_crop_script_parameters: BlenderCropScriptParameters
+) -> Dict[str, Union[int, List[float], bool]]:
+    return dict(
+        resolution=[
+            blender_crop_script_parameters.resolution.x_dimension,
+            blender_crop_script_parameters.resolution.y_dimension,
+        ],
+        borders_x=[
+            float(blender_crop_script_parameters.borders_x.minimum_border_edge),
+            float(blender_crop_script_parameters.borders_x.maximum_border_edge),
```
Comment (cameel) [inline concent_api/conductor/utils.py]: Conversion to `float` is lossy. You're potentially losing precision.

You should either keep Decimals here (if Blender can handle them) or just use `float` in the model (if it can't).

If it's because you can't use Decimals in a RabbitMQ task definition, then just pass them around as strings.

File: concent_api/core/tests/test_unit_validation.py [hunk 21]
```diff
@@ -441,3 +447,212 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+    @pytest.mark.parametrize(
+        'use_compositing', [
+            True,
+            False
+        ]
+    )  # pylint: disable=no-self-use
+    def test_that_valid_use_compositing_value_doesnt_raise_exception(self, use_compositing):
+        try:
+            validate_use_compositing(use_compositing)
+        except Exception as exception:  # pylint: disable=broad-except
```
Comment (cameel) [inline concent_api/core/tests/test_unit_validation.py]: This is too broad. You should expect a specific exception type here.

File: concent_api/core/tests/test_unit_validation.py [hunk 22]
```diff
@@ -441,3 +447,212 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+        ['borders_x', 'borders_y'], [
+            [[0.0, 1.0], [0.0, 1.0]],
+            [[0.71830, 1.0], [0.9, 1.0]],
+        ]
+    )
+    def test_that_valid_crops_value_doesnt_raise_exception(self, borders_x, borders_y):
+        try:
+            validate_crops([self.create_crop_dict(borders_x, borders_y)])
+        except Exception as exception:  # pylint: disable=broad-except
+            assert False, f"Unexpected exception has been raised: {str(exception)}"
```
Comment (cameel) [inline concent_api/core/tests/test_unit_validation.py]: This is too broad.

File: concent_api/conductor/models.py (hunk 18 above)
Comment (kbeker) [reply concent_api/conductor/models.py]: Hmm, but this would extend `BlenderSubtaskDefinition` table too much I think that script parameters could be independent table. I changed that now `BlenderCropScriptParameters` is merged all in one, but I'd stay with this parameters in one separate table instead keeping all of them i `BlenderSubtaskDefinition` model

File: concent_api/conductor/utils.py (hunk 20 above)
Comment (kbeker) [reply concent_api/conductor/utils.py]: Changed conversion to strings

File: concent_api/core/tests/test_unit_validation.py (hunk 21 above)
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: I think that we should catch here all exception because we assume that this test should pass without any error, but as you suggested I changed to `ConcentValidationError`

File: concent_api/core/tests/test_unit_validation.py (hunk 22 above)
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: Done

File: concent_api/conductor/service.py [hunk 23]
```diff
@@ -113,3 +119,38 @@ def store_frames(
 
 def filter_frames_by_blender_subtask_definition(blender_subtask_definition: BlenderSubtaskDefinition) -> list:
     return list(Frame.objects.filter(blender_subtask_definition=blender_subtask_definition).values_list('number', flat=True))
+
+
+def store_blender_crop_script_parameters(blender_parameters: Any) -> BlenderCropScriptParameters:
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/conductor/service.py]: I changed to `_store_blender_crop_script_parameters`

File: concent_api/core/tests/test_unit_validation.py [hunk 24]
```diff
@@ -441,3 +447,212 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+    @pytest.mark.parametrize(
+        'resolution', [
+            [1, 1],
+            [100000000, 100000000],
+        ]
+    )  # pylint: disable=no-self-use
+    def test_that_valid_resolution_value_doesnt_raise_exception(self, resolution):
+        try:
+            validate_resolution(resolution)
+        except Exception as exception:  # pylint: disable=broad-except
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: It is not redundant :) This is here because if anything goes wrong it will catch this exception

File: concent_api/core/tests/test_unit_validation.py [hunk 25]
```diff
@@ -441,3 +447,212 @@ def test_that_unsupported_format_raises_concent_validation_error(self):  # pylin
...
+            ['100', 100],
+            [],
+            ['400x800'],
+            [71830.23, 1000],
+        ]
+    )  # pylint: disable=no-self-use
+    def test_that_invalid_resolution_value_raise_exception(self, resolution):
+        with pytest.raises(ConcentValidationError) as exception_wrapper:
+            validate_resolution(resolution)
+        assert_that(exception_wrapper.value.error_code).is_equal_to(ErrorCode.MESSAGE_INVALID)
```
[the remark this answers is not in the document]
Comment (kbeker) [reply concent_api/core/tests/test_unit_validation.py]: I was thinking about it, but until Concent's `error_messages` is not standardized and all error_messages are string then it won't be effective to check it because it might change. Also if we want to check it this test will be illegible:) The most important thing in this test is to check if all wrong values will raise exception. If we want to check that all validation has been tested we could run tests with coverage and then we can see that everything has been tested :)

File: concent_api/conductor/service.py (hunk 19 above)
Comment (kbeker) [reply concent_api/conductor/service.py]: I changed https://github.com/golemfactory/concent/pull/1128#discussion_r269887568.
